### PR TITLE
cpp/incorrect-string-type-conversion false positive fixes

### PR DESCRIPTION
**Pull request by bdrodes**
Fixes two false positives with cpp/incorrect-string-type-conversion: when conversion is on a 'byte' type and conversion on new array expressions.

### CODE REVIEW

File: cpp/ql/src/Security/CWE/CWE-704/WcharCharConversion.ql [hunk 1]
```diff
@@ -18,13 +18,31 @@ class WideCharPointerType extends PointerType {
   WideCharPointerType() { this.getBaseType() instanceof WideCharType }
 }
 
+/**
+ * types that may also be `CharPointerType`, but that are likely used as arbitrary buffers
```
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-704/WcharCharConversion.ql:22]:
```suggestion
 * A type that may also be `CharPointerType`, but that are likely used as arbitrary buffers.
```

**Review by geoffw0 — COMMENTED** [review]
LGTM, but it would be nice to have a test for each new case , particularly the 'casting from new' one.

**Review by geoffw0 — COMMENTED** [review]
Thanks for adding the test cases, this change looks good to me.

CI wants you to add a change note.  I think it's right, a brief note saying that this query has been refined to remove likely false positive cases could be helpful to other users of the query.

**Comment by bdrodes** [comment]
> Thanks for adding the test cases, this change looks good to me.
> 
> CI wants you to add a change note. I think it's right, a brief note saying that this query has been refined to remove likely false positive cases could be helpful to other users of the query.

Can you remind mee where there is supposed to go?

**Comment by geoffw0** [comment]
Query change notes go in `ql\cpp\ql\src\change-notes` (library change notes go in a slightly different location).  The category for this one is probably "minorAnalysis".

**Comment by geoffw0** [comment]
Oh and the result changes on the DCA run LGTM - all appear to be converting buffers rather than strings, or converting buffers _into_ strings, which are exactly the kind of results this PR aims to hide.

**Comment by bdrodes** [comment]
> Query change notes go in `ql\cpp\ql\src\change-notes` (library change notes go in a slightly different location). The category for this one is probably "minorAnalysis".

Got it. Uploaded.

File: cpp/ql/src/change-notes/2024-01-29-false_positive_incorrect_string_type_conversion.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* Corrected 2 false positive with `cpp/incorrect-string-type-conversion`: conversion of byte arrays to w_char and new array allocations converted to w_char.
```
Comment (geoffw0) [inline cpp/ql/src/change-notes/2024-01-29-false_positive_incorrect_string_type_conversion.md:4]:
```suggestion
* Corrected 2 false positive with `cpp/incorrect-string-type-conversion`: conversion of byte arrays to wchar and new array allocations converted to wchar.
```

**Review by geoffw0 — APPROVED** [review]
Sorry for all the nitpicks.  LGTM.

**Comment by bdrodes** [comment]
> Sorry for all the nitpicks. LGTM.

No worries. Out of curiosity, once a PR is approved, who gives the final go ahead to merge?

**Comment by MathiasVP** [comment]
> No worries. Out of curiosity, once a PR is approved, who gives the final go ahead to merge?

We don't have a very strict requirement here. As soon as the PR has been approved it should generally be fine to hit merge.

**Pull request merged**